### PR TITLE
Implemented the HeapArray Class for Utils

### PR DESCRIPTION
**Pull request by Raelr**
Implemented a HeapArray class. The HeapArray is a heap-allocated array which tracks all inputs the array receives. The class acts as a normal array, however it has one constraint - you can only access elements which you have explicitly inserted into the array. Any access of indices which were not inserted explicitly are considered 'garbage' data, and will cause an assertion failure when accessed.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -66,13 +66,13 @@ $(coreLib): buildFlags
...
 
-$(testApp): buildFlags $(utilsLib) $(coreLib)
+$(testApp): buildFlags $(utilsLib) $(coreLib) $(renderLib)
 	"$(MAKE)" -C $(testsDir) CXXFLAGS="$(CXXFLAGS)"
 
 $(exampleGameApp): buildFlags $(utilsLib) $(coreLib)
 	"$(MAKE)" -C $(examplesDir)/game CXXFLAGS="$(CXXFLAGS)"
 
-$(exampleRenderApp): buildFlags $(renderLib)
+$(exampleRenderApp): buildFlags $(utilsLib) $(renderLib)
```
Comment (J-Mo63) [inline Makefile:75]: Again, unless you're testing `renderlib`, don't add it as a dip of `testApp`. Did you ever figure out why `utilslib` is needed here and if we can remove it?

File: engine/render/renderer/Core.h [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #include <cstdint>
 #include <iostream>
 
+#include "../../utils/HeapArray.h"
 #include "utils/Array.h"
 #include "utils/Hash.h"
 #include "utils/StackArray.h"
```
Comment (J-Mo63) [inline engine/render/renderer/Core.h:19]: utils classes should be brought in with without relative pathing, such as `#include <utils/HeapArray.h>` etc.

File: tests/Makefile [hunk 3]
```diff
@@ -17,7 +17,7 @@ testBuildDir := $(testBinDir)/build
 
 # Set build vars
 utestIncludeDir := $(vendorDir)/include/utest
-linkFlags += -l core -l utils
+linkFlags += -l core -l utils -lrender
```
Comment (J-Mo63) [inline tests/Makefile:20]: why is this necessary?

File: tests/core/test_EntitySystem.cpp [hunk 4]
```diff
@@ -35,6 +36,17 @@ UTEST(test_EntitySystem, AddEntities)
     // The storage should contain the new entities after registration
     system.RegisterEntities();
     ASSERT_EQ(3, system.GetEntities().size());
+
+    // TODO: Change the Entity constructor to be able to take in string temporaries.
+    String e4Name = "e4";
+    String e5Name = "e5";
+
+    // TODO: If this takes in a string temporary it will break Linux.
```
Comment (J-Mo63) [inline tests/core/test_EntitySystem.cpp:44]: Nice find

File: engine/utils/HeapArray.h [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <initializer_list>
+#include <utility>
+
+typedef const size_t sizet;
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:20]: This is quite unnecessary

File: engine/utils/HeapArray.h [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+ * @tparam T the type of the pointer.
+ * @param ptr the pointer to be reallocated.
+ * @param dataSize the size of the new memory chunk.
+ * @return the new reallocated pointer.
+ */
+template<typename T>
+static inline T* Reallocate(T* ptr, const size_t& dataSize)
+{
+    return static_cast<T*>(realloc(ptr, dataSize));
+}
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:50]: These functions should be moved to the source file

File: engine/utils/HeapArray.h [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+        CopyData(other.data, sizeof(T) * size);
+        CopyMasks(other.stateMasks, BYTE_MASK_SIZE * byteCount);
+    }
+
+    // Tracker Variables
+    size_t size {0};
+    size_t count {0};
+
+    // Pointers
+    uint8_t* stateMasks {nullptr};
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:729]: Why does this need to be a pointer? A `uint8_t` is the same size as a pointer, so there's no space savings and it poses an additional point of error to cause a memory leak.

File: engine/utils/HeapArray.h [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+
+    /**
+     * @brief Allocates our raw pointers to the specified values.
+     * @param arraySize the size to allocate `data` to.
+     * @param bytesCount the number of bytes to assign for states.
+     */
+    void AllocateMemory(const size_t& arraySize, const size_t& bytesCount)
+    {
+        data = Allocate<T>(sizeof(T) * arraySize);
+        stateMasks = Allocate<uint8_t>(BYTE_MASK_SIZE * bytesCount);
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:642]: Possible memory leak

File: engine/utils/HeapArray.h [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+        size_t size;
+    };
+
+    /**
+     * @brief An empty constructor. This is invoked when no parameters are passed to the
+     * constructor. Initialises the array with a size of zero. Does not allocate any space on the
+     * heap. Constructing an array this way requires either an explicit resizing or a new creation
+     * to add a size. Any usage of the Get() or subscript operator will cause an error.
+     */
+    HeapArray() : HeapArray(0, 0, 0, nullptr, nullptr) {}
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:209]: Is there any real use-case for this constructor?

File: engine/utils/HeapArray.h [hunk 10]
```diff
@@ -0,0 +1,735 @@
...
+    {
+        size = list.size();
+
+        data = Allocate<T>(sizeof(T) * list.size());
+
+        size_t newByteCount = CalculateByteMaskCount(list.size());
+
+        stateMasks = Allocate<uint8_t>(BYTE_MASK_SIZE * newByteCount);
+
+        memset(stateMasks, 0, (BYTE_MASK_SIZE * newByteCount));
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:233]: This pattern of allocation and memset would not be an issue if it weren't a pointer.

File: engine/utils/HeapArray.h [hunk 11]
```diff
@@ -0,0 +1,735 @@
...
+
+        for (auto& element : list) Insert(count, element);
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:235]: May be faster to just copy `list.data()` and then check what you have after the fact?

**Review by J-Mo63 — COMMENTED** [review]
This is a partial review. I'll need to come back and continue later.

File: engine/utils/HeapArray.h (hunk 7 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:729]: Because we aren't storing a single 8-bit unsigned integer. We're storing multiple 8-bit unsigned integers for every 8 elements in the array. The pointer indicates that this is an array value.

File: engine/utils/HeapArray.h (hunk 11 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:235]: unfortunately `list.data()` is not publicly accessible in a `std::initializer_list`. Trust me I've tried.

File: engine/utils/HeapArray.h (hunk 10 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:233]: Again, it has to be a pointer because we're storing an array of bitmasks, not a single bitmask.

File: engine/utils/HeapArray.h (hunk 9 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:209]: If you want to create an empty HeapArray which will be used later. I use this pattern a few times in the renderer.

File: tests/Makefile (hunk 3 above)
Comment (Raelr) [reply tests/Makefile:20]: It isn't. This is an artifact from when I was experimenting with having this in `render`

File: engine/render/renderer/Core.h (hunk 2 above)
Comment (Raelr) [reply engine/render/renderer/Core.h:19]: Good catch.

File: Makefile (hunk 1 above)
Comment (Raelr) [reply Makefile:75]: Yes, this is again an artifact from some previous coding. I'll get rid of this dependency.

File: engine/utils/HeapArray.h (hunk 7 above)
Comment (J-Mo63) [reply engine/utils/HeapArray.h:729]: Ah, then rename this to something like `uint8_t* stateMaskBitfield` to be more accurate about what it is. Alternatively you could use `std::vector<bool>` since that will use a bitfield under the hood and manage it for you.

File: engine/utils/HeapArray.h (hunk 10 above)
Comment (J-Mo63) [reply engine/utils/HeapArray.h:233]: Then roll these two into a function if it's going to be done multiple times across the class.

File: engine/utils/HeapArray.h (hunk 11 above)
Comment (J-Mo63) [reply engine/utils/HeapArray.h:235]: It's not a member function, it's free floating, right?

Comment (Raelr) [reply engine/utils/HeapArray.h:235]: no, it's straight up inaccessible

Comment (J-Mo63) [reply engine/utils/HeapArray.h:235]: This compiles just fine for me:
```cpp
#include <iostream>
#include <initializer_list>

void test(std::initializer_list<int> list) {
  const int* i = std::data(list);
  std::cout << "Hello World! " << *(i+3) << *(i+2) << *(i+1) << *(i);
}

int main() {
  test({7, 3, 3, 1});
  return 0;
}
```

File: Makefile (hunk 1 above)
Comment (Raelr) [reply Makefile]: Fixed.

File: engine/render/renderer/Core.h (hunk 2 above)
Comment (Raelr) [reply engine/render/renderer/Core.h]: Changed.

File: tests/Makefile (hunk 3 above)
Comment (Raelr) [reply tests/Makefile]: Removed it.

File: engine/utils/HeapArray.h (hunk 5 above)
Comment (Raelr) [reply engine/utils/HeapArray.h]: you are correct - I remember this was leftover from a small experiment.

File: engine/utils/HeapArray.h (hunk 6 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:48]: I don't understand what you mean by this. The fact that these are templated means that they can't be moved to a `.cpp` file. They have to live in the header.

File: engine/utils/HeapArray.h (hunk 7 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:727]: Changed the name.

File: engine/utils/HeapArray.h (hunk 11 above)
Comment (Raelr) [reply engine/utils/HeapArray.h]: Ahhhhhh, fair enough. I stand corrected. I've edited the solution accordingly.

File: engine/utils/HeapArray.h (hunk 10 above)
Comment (Raelr) [reply engine/utils/HeapArray.h]: sure thang - I'll get to this over the weekend since I've run out of time.

Comment (Raelr) [reply engine/utils/HeapArray.h]: Just had a look - this pattern is not repeated anywhere else in the class.

File: engine/utils/HeapArray.h (hunk 8 above)
Comment (Raelr) [reply engine/utils/HeapArray.h]: Had a look into this. AllocateMemory is being used in two locations:

```cpp
HeapArray(const HeapArray<T>& other) : size {other.size}, count {other.count}
    {
        if (other.data == nullptr && other.size == 0) return;

        auto stateMaskCount = CalculateByteMaskCount(other.size);

        AllocateMemory(size, stateMaskCount);

        CopyData(other.data, sizeof(T) * size);
        CopyMasks(other.stateMaskBitfield, BYTE_MASK_SIZE * stateMaskCount);

        count = other.Count();
    }
```

and 

```cpp
    void Copy(const HeapArray<T>& other)
    {
        if (other.size == 0 && other.data == nullptr) return;

        count = other.count;
        size = other.size;

        auto byteCount = CalculateByteMaskCount(other.size);

        AllocateMemory(other.size, byteCount);

        CopyData(other.data, sizeof(T) * size);
        CopyMasks(other.stateMaskBitfield, BYTE_MASK_SIZE * byteCount);
    }
```

In the first case, the array will never be pre-allocated since its a copy constructor. There's no danger of a memory leak in that scenario.

In the second case, there is a bit of a risk, since the array could have some pre-allocated data. I've added a check which will de-allocate the data accordingly: 

```cpp
// If the array is currently being used then de-allocate the array and reallocate it.
        if (data != nullptr) Destroy();

        AllocateMemory(other.size, byteCount);
```

Comment (Raelr) [reply engine/utils/HeapArray.h]: Correction - I've changed it to reallocate memory instead. That way we can allocate memory if the array hasn't been allocated and resize it if it's a different size.

File: engine/utils/HeapArray.h (hunk 6 above)
Comment (J-Mo63) [reply engine/utils/HeapArray.h:48]: Remember to move this to another namespace so it doesn't collide

File: engine/utils/HeapArray.h [hunk 12]
```diff
@@ -0,0 +1,764 @@
...
+
+// HeapArray
+
+/**
+ * @brief The HeapArray is a managed heap-allocated array type. The HeapArray is intended to be
+ * used when a container is needed who's size is not known at compile time, but is required to be
+ * constant once a size is provided.
+ *
+ * The HeapArray tracks all data that's been inserted and removed from the array. All data inserted
+ * into the array will have its indexed marked as active. Accessing an index which has not been
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:58]:
```suggestion
 * into the array will have its indices marked as active. Accessing an index which has not been
```

File: engine/utils/HeapArray.h [hunk 13]
```diff
@@ -0,0 +1,764 @@
...
+                          const size_t& newSize) :
+            pointer {valuePointer},
+            maskPointer {newMaskPointer},
+            index {newIndex},
+            size {newSize}
+        {}
+
+        /**
+         * @brief The iterator prefix increment operator. This operator will increment the pointer
+         * as many times as needed to find an active element. As such, it will not return unactive
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:99]:
```suggestion
         * as many times as needed to find an active element. As such, it will not return inactive
```

File: engine/utils/HeapArray.h [hunk 14]
```diff
@@ -0,0 +1,764 @@
...
+        {
+            return pointer != other.pointer;
+        }
+
+    private:
+
+        T* pointer;
+        uint8_t* maskPointer;
+        size_t index;
+        size_t size;
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:198]: This can just be a pointer to the heap array itself, no?

File: engine/utils/HeapArray.h [hunk 15]
```diff
@@ -0,0 +1,764 @@
...
+     * @param other the array to be moved.
+     */
+    HeapArray(HeapArray<T>&& other) noexcept :
+        size {std::move(other.size)},
+        count {std::move(other.count)},
+        stateMaskBitfield {std::move(other.stateMaskBitfield)},
+        data {std::move(other.data)}
+    {
+        other.ResetValues();
+    }
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:281]: This could use swap

File: engine/utils/HeapArray.h [hunk 16]
```diff
@@ -0,0 +1,764 @@
...
+        return Iterator(data, stateMaskBitfield, 0, size);
+    }
+
+    /**
+     * @brief Creates an iterator pointing to the end of the HeapArray.
+     * @return an Iterator set to the end of the array.
+     */
+    Iterator end()
+    {
+        return Iterator(data + size, stateMaskBitfield, size, size);
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:534]:
```suggestion
        return {data + size, stateMaskBitfield, size, size};
```

File: engine/utils/HeapArray.h [hunk 17]
```diff
@@ -0,0 +1,764 @@
...
+    }
+
+private:
+
+    // Class constants
+
+    // The number of states we can fit per segment (1 byte = 8 states)
+    static constexpr uint8_t BYTE_SIZE_IN_BITS = 8;
+    // The default object size for each segment (1 byte)
+    static constexpr uint8_t BYTE_MASK_SIZE = sizeof(uint8_t);
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:562]:
```suggestion
    static constexpr uint8_t BITMASK_SIZE = sizeof(uint8_t);
```

File: engine/utils/HeapArray.h [hunk 18]
```diff
@@ -0,0 +1,764 @@
...
+     * @param index the index to evaluate.
+     */
+    void CheckIfActive(const size_t& index)
+    {
+        assert(Active(index) && "Element does not exist in the array!");
+    }
+
+    size_t CalculateByteMaskCount(const size_t& arraySize)
+    {
+        return ((arraySize / BYTE_SIZE_IN_BITS) + 1);
```
Comment (J-Mo63) [inline engine/utils/HeapArray.h:648]:
```suggestion
        return (arraySize / BYTE_SIZE_IN_BITS) + 1;
```

File: tests/utils/test_HeapArray.cpp [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+    Siege::Utils::HeapArray<uint32_t> array({1});
+
+    uint32_t expectedResults[] = {1};
+
+    size_t index = 0;
+    for (auto& element : array)
+    {
+        ASSERT_EQ(expectedResults[index++], element);
+    }
+}
```
Comment (J-Mo63) [inline tests/utils/test_HeapArray.cpp:329]: Add case for 0 elems?

**Review by J-Mo63 — APPROVED** [review]
Looks good. If you can do some cleanup and rerun it, good to merge.

File: engine/utils/HeapArray.h (hunk 13 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:99]: Done

File: engine/utils/HeapArray.h (hunk 12 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:58]: Done.

File: engine/utils/HeapArray.h (hunk 14 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:198]: Done. I've also changed the iterators to work like UE5 iterators.

File: engine/utils/HeapArray.h (hunk 15 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:281]: Done.

File: engine/utils/HeapArray.h (hunk 17 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:562]: Done.

File: engine/utils/HeapArray.h (hunk 18 above)
Comment (Raelr) [reply engine/utils/HeapArray.h:648]: Done.

File: tests/utils/test_HeapArray.cpp (hunk 19 above)
Comment (Raelr) [reply tests/utils/test_HeapArray.cpp:329]: Added the case!
